### PR TITLE
Branch add Gender and Weight class (Can merge)

### PR DESCRIPTION
**Pull request by dfordarius**
The current version does not fulfill the requirements of FitBook

Lets add Gender and Weight classes so that Person's information such as gender and weight can be stored as well. 

Added: 
- Gender Class
- Weight Class

Updated: 
- Person.java to include weight and gender
- AddCommand to include weight and gender
- EditCommand to include weight and gender
- Testing

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 1]
```diff
@@ -95,6 +97,34 @@ public static Email parseEmail(String email) throws ParseException {
...
     }
+    /**
+     * Parses a {@code String Weight} into an {@code Weight}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code weight} is invalid.
+     */
+    public static Weight parseWeight(String weight) throws ParseException {
+        requireNonNull(weight);
+        String trimmedAddress = weight.trim();
```
Comment (jundatan) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:108]: Better to rename String to trimmedWeight.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -95,6 +97,34 @@ public static Email parseEmail(String email) throws ParseException {
...
+    }
+    /**
+     * Parses a {@code String Gender} into an {@code Gender}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code gender} is invalid.
+     */
+    public static Gender parseGender(String gender) throws ParseException {
+        requireNonNull(gender);
+        String trimmedAddress = gender.trim();
```
Comment (jundatan) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:122]: Better to rename String to trimmedGender.

File: src/main/java/seedu/address/model/client/Gender.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package seedu.address.model.client;
+
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Client's phone number in the address book.
```
Comment (jundatan) [inline src/main/java/seedu/address/model/client/Gender.java:8]: Comment not in line with the class and should be in FitBook.

File: src/main/java/seedu/address/model/client/Gender.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Gender should either be M or F (not case sensitive)";
+    public static final String VALIDATION_REGEX = "(M|F|m|f)";
+    public final String value;
+
+    /**
+     * Constructs a {@code Weight}.
+     *
+     * @param gender A valid phone number.
```
Comment (jundatan) [inline src/main/java/seedu/address/model/client/Gender.java:21]: @param needs to change to gender.

File: src/main/java/seedu/address/model/client/Weight.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
+package seedu.address.model.client;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Client's phone number in the address book.
```
Comment (jundatan) [inline src/main/java/seedu/address/model/client/Weight.java:7]: Comment not in line with the class and should be in FitBook.

File: src/main/java/seedu/address/model/client/Weight.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Weight should only contain positive numbers and cannot be 0";
+    public static final String VALIDATION_REGEX = "^[1-9]\\d*.\\d*|0.\\d*[1-9]\\d*$";
+    public final String value;
+
+    /**
+     * Constructs a {@code Weight}.
+     *
+     * @param weight A valid phone number.
```
Comment (jundatan) [inline src/main/java/seedu/address/model/client/Weight.java:21]: @param needs to change to gender.

File: src/main/java/seedu/address/model/util/SampleDataUtil.java [hunk 7]
```diff
@@ -10,34 +10,37 @@
...
 import seedu.address.model.tag.Tag;
 
 /**
  * Contains utility methods for populating {@code FitBook} with sample data.
  */
 public class SampleDataUtil {
+
     public static Client[] getSampleClients() {
-        return new Client[] {
+        return new Client [] {
```
Comment (jundatan) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java:25]: Dont think there should be a space between Client and [].

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java [hunk 8]
```diff
@@ -29,7 +31,12 @@ class JsonAdaptedClient {
     private final String phone;
     private final String email;
     private final String address;
+
+    private final String weight;
+    private final String gender;
```
Comment (jundatan) [inline src/main/java/seedu/address/storage/JsonAdaptedClient.java:36]: Think these String declarations can be group together.

File: src/main/java/seedu/address/storage/JsonAdaptedClient.java [hunk 9]
```diff
@@ -29,7 +31,12 @@ class JsonAdaptedClient {
...
+
     private final List<JsonAdaptedAppointment> appointments = new ArrayList<>();
+
     private final List<JsonAdaptedTag> tagged = new ArrayList<>();
```
Comment (jundatan) [inline src/main/java/seedu/address/storage/JsonAdaptedClient.java:40]: These String declarations can be group together as they are non compulsory.

File: src/test/data/JsonFitBookStorageTest/invalidAndValidClientFitBook.json [hunk 10]
```diff
@@ -3,11 +3,13 @@
     "name": "Valid Client",
     "phone": "9482424",
     "email": "hans@example.com",
-    "address": "4th street"
```
Comment (jundatan) [inline src/test/data/JsonFitBookStorageTest/invalidAndValidClientFitBook.json:6]: I think the address part should not be removed.

File: src/test/data/JsonFitBookStorageTest/invalidAndValidClientFitBook.json [hunk 11]
```diff
@@ -3,11 +3,13 @@
...
+    "weight": "23",
+    "gender": "m"
   }, {
     "name": "Client With Invalid Phone Field",
     "phone": "948asdf2424",
     "email": "hans@example.com",
-    "address": "4th street"
```
Comment (jundatan) [inline src/test/data/JsonFitBookStorageTest/invalidAndValidClientFitBook.json:11]: I think the address part should not be removed.

File: src/test/data/JsonSerializableFitBookTest/typicalClientsFitBook.json [hunk 12]
```diff
@@ -5,49 +5,66 @@
     "phone" : "94351253",
     "email" : "alice@example.com",
     "address" : "123, Jurong West Ave 6, #08-111",
-    "tagged" : [ "friends" ],
-    "appointments": [ "13-02-1999" ]
+    "appointments": [ "13-02-2020", "14-01-2021" ],
```
Comment (jundatan) [inline src/test/data/JsonSerializableFitBookTest/typicalClientsFitBook.json:8]: Just want to ask why is this changed? hahaha

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 13]
```diff
@@ -56,45 +60,52 @@ public void parse_allFieldsPresent_success() {
...
 
         // multiple appointments - all accepted
         Client expectedClientMultipleAppointments = new ClientBuilder(BOB).withTags(VALID_TAG_FRIEND, VALID_TAG_HUSBAND)
                 .withAppointments(VALID_APPOINTMENT_DATE_ONE, VALID_APPOINTMENT_DATE_TWO).build();
         assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_BOB
-                        + TAG_DESC_HUSBAND + TAG_DESC_FRIEND + APPOINTMENT_DESC_DATE_ONE + APPOINTMENT_DESC_DATE_TWO,
+                        + GENDER_DESC_BOB + WEIGHT_DESC_BOB + APPOINTMENT_DESC_DATE_ONE + APPOINTMENT_DESC_DATE_TWO
+                        + TAG_DESC_HUSBAND + TAG_DESC_FRIEND,
                 new AddCommand(expectedClientMultipleAppointments));
+
```
Comment (jundatan) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:100]: Think is better to remove this empty line.

File: src/test/java/seedu/address/logic/parser/AddCommandParserTest.java [hunk 14]
```diff
@@ -126,35 +137,41 @@ public void parse_compulsoryFieldMissing_failure() {
...
         // invalid tag
         assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_BOB
-                + INVALID_TAG_DESC + VALID_TAG_FRIEND + APPOINTMENT_DESC_DATE_ONE, Tag.MESSAGE_CONSTRAINTS);
+                + WEIGHT_DESC_BOB + GENDER_DESC_BOB + INVALID_TAG_DESC + VALID_TAG_FRIEND, Tag.MESSAGE_CONSTRAINTS);
 
         // invalid appointment
         assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + ADDRESS_DESC_BOB
-                + TAG_DESC_HUSBAND + INVALID_APPOINTMENT_DESC, Appointment.MESSAGE_CONSTRAINTS);
+                + WEIGHT_DESC_BOB + GENDER_DESC_BOB + TAG_DESC_HUSBAND + INVALID_APPOINTMENT_DESC,
+                Appointment.MESSAGE_CONSTRAINTS);
```
Comment (jundatan) [inline src/test/java/seedu/address/logic/parser/AddCommandParserTest.java:166]: Maybe will be better to add extra invalid tests for gender and weight?

File: src/test/java/seedu/address/model/client/ClientTest.java [hunk 15]
```diff
@@ -87,8 +88,16 @@ public void equals() {
...
+        //different weight -> returns false
+        editedAlice = new ClientBuilder(ALICE).withWeight(VALID_WEIGHT_BOB).build();
+        assertFalse(ALICE.equals(editedAlice));
+
+        //different gender -> returns false
+        editedAlice = new ClientBuilder(ALICE).withGender("f").build();
+
         // different appointments -> returns false
         editedAlice = new ClientBuilder(ALICE).withAppointments(VALID_APPOINTMENT_DATE_TWO).build();
+
```
Comment (jundatan) [inline src/test/java/seedu/address/model/client/ClientTest.java:100]: better to remove this empty line.

File: src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java [hunk 16]
```diff
@@ -9,12 +9,15 @@
...
+import seedu.address.model.client.Gender;
 import seedu.address.model.client.Name;
 import seedu.address.model.client.Phone;
+import seedu.address.model.client.Weight;
 import seedu.address.model.tag.Tag;
 
+
 /**
- * A utility class to help with building EditClientDescriptor objects.
+ * A utility class to help with building EditPersonDescriptor objects.
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java:20]: Needs to change to EditClientDescriptor.

File: src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java [hunk 17]
```diff
@@ -29,52 +32,70 @@ public EditClientDescriptorBuilder(EditClientDescriptor descriptor) {
     }
 
     /**
-     * Returns an {@code EditClientDescriptor} with fields containing {@code client}'s details
+     * Returns an {@code EditPersonDescriptor} with fields containing {@code client}'s details
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java:35]: Needs to change to EditClientDescriptor.

File: src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java [hunk 18]
```diff
@@ -29,52 +32,70 @@ public EditClientDescriptorBuilder(EditClientDescriptor descriptor) {
...
         descriptor.setAddress(client.getAddress());
+        descriptor.setWeight(client.getWeight());
+        descriptor.setGender(client.getGender());
         descriptor.setAppointments(client.getAppointments());
         descriptor.setTags(client.getTags());
     }
 
     /**
-     * Sets the {@code Name} of the {@code EditClientDescriptor} that we are building.
+     * Sets the {@code Name} of the {@code EditPersonDescriptor} that we are building.
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java:50]: Needs to change to EditClientDescriptor.

File: src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java [hunk 19]
```diff
@@ -29,52 +32,70 @@ public EditClientDescriptorBuilder(EditClientDescriptor descriptor) {
...
      */
     public EditClientDescriptorBuilder withName(String name) {
         descriptor.setName(new Name(name));
         return this;
     }
 
     /**
-     * Sets the {@code Phone} of the {@code EditClientDescriptor} that we are building.
+     * Sets the {@code Phone} of the {@code EditPersonDescriptor} that we are building.
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java:58]: Needs to change to EditClientDescriptor.

File: src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java [hunk 20]
```diff
@@ -29,52 +32,70 @@ public EditClientDescriptorBuilder(EditClientDescriptor descriptor) {
...
      */
     public EditClientDescriptorBuilder withPhone(String phone) {
         descriptor.setPhone(new Phone(phone));
         return this;
     }
 
     /**
-     * Sets the {@code Email} of the {@code EditClientDescriptor} that we are building.
+     * Sets the {@code Email} of the {@code EditPersonDescriptor} that we are building.
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java:66]: Needs to change to EditClientDescriptor.

File: src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java [hunk 21]
```diff
@@ -29,52 +32,70 @@ public EditClientDescriptorBuilder(EditClientDescriptor descriptor) {
...
      */
     public EditClientDescriptorBuilder withEmail(String email) {
         descriptor.setEmail(new Email(email));
         return this;
     }
 
     /**
-     * Sets the {@code Address} of the {@code EditClientDescriptor} that we are building.
+     * Sets the {@code Address} of the {@code EditPersonDescriptor} that we are building.
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java:74]: Needs to change to EditClientDescriptor.

File: src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java [hunk 22]
```diff
@@ -29,52 +32,70 @@ public EditClientDescriptorBuilder(EditClientDescriptor descriptor) {
...
      */
     public EditClientDescriptorBuilder withAddress(String address) {
         descriptor.setAddress(new Address(address));
         return this;
     }
 
     /**
-     * Parses the {@code tags} into a {@code Set<Tag>} and set it to the {@code EditClientDescriptor}
+     * Sets the {@code weight} of the {@code EditPersonDescriptor} that we are building.
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java:82]: Needs to change to EditClientDescriptor.

File: src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java [hunk 23]
```diff
@@ -29,52 +32,70 @@ public EditClientDescriptorBuilder(EditClientDescriptor descriptor) {
...
+     */
+    public EditClientDescriptorBuilder withWeight(String weight) {
+        descriptor.setWeight(new Weight(weight));
+        return this;
+    }
+
+    /**
+     * Sets the {@code gender} of the {@code EditPersonDescriptor} that we are building.
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java:90]: Needs to change to EditClientDescriptor.

File: src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java [hunk 24]
```diff
@@ -29,52 +32,70 @@ public EditClientDescriptorBuilder(EditClientDescriptor descriptor) {
...
+     */
+    public EditClientDescriptorBuilder withGender(String gender) {
+        descriptor.setGender(new Gender(gender));
+        return this;
+    }
+
+    /**
+     * Parses the {@code tags} into a {@code Set<Tag>} and set it to the {@code EditPersonDescriptor}
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/EditClientDescriptorBuilder.java:98]: Needs to change to EditClientDescriptor.

File: src/test/java/seedu/address/testutil/TypicalClients.java [hunk 25]
```diff
@@ -27,38 +30,46 @@ public class TypicalClients {
...
 
     // Manually added - Person's details found in {@code CommandTestUtil}
     public static final Client AMY = new ClientBuilder().withName(VALID_NAME_AMY).withPhone(VALID_PHONE_AMY)
-            .withEmail(VALID_EMAIL_AMY).withAddress(VALID_ADDRESS_AMY).withTags(VALID_TAG_FRIEND)
-            .withAppointments(VALID_APPOINTMENT_DATE_ONE).build();
+            .withEmail(VALID_EMAIL_AMY).withAddress(VALID_ADDRESS_AMY)
+            .withWeight(VALID_WEIGHT_AMY).withGender(VALID_GENDER_AMY).withAddress(VALID_ADDRESS_AMY)
+            .withTags(VALID_TAG_FRIEND).withAppointments(VALID_APPOINTMENT_DATE_ONE).build();
     public static final Client BOB = new ClientBuilder().withName(VALID_NAME_BOB).withPhone(VALID_PHONE_BOB)
-            .withEmail(VALID_EMAIL_BOB).withAddress(VALID_ADDRESS_BOB).withTags(VALID_TAG_HUSBAND, VALID_TAG_FRIEND)
```
Comment (jundatan) [inline src/test/java/seedu/address/testutil/TypicalClients.java:59]: Want to ask what is the reason to remove the tags for Bob?

Comment (dfordarius) [reply src/test/java/seedu/address/testutil/TypicalClients.java:59]: Used 'bob' in testing as a typical client with no tags allocated. To support testing for adding/editing tags. I will just add back the tags for consistency, and change the testing

Comment (jundatan) [reply src/test/java/seedu/address/testutil/TypicalClients.java:59]: Okay. If you find it is better to change it, then we can keep it that way.

File: src/test/java/seedu/address/model/client/ClientTest.java [hunk 26]
```diff
@@ -87,6 +88,13 @@ public void equals() {
         editedAlice = new ClientBuilder(ALICE).withAddress(VALID_ADDRESS_BOB).build();
         assertFalse(ALICE.equals(editedAlice));
 
+        //different weight -> returns false
+        editedAlice = new ClientBuilder(ALICE).withWeight(VALID_WEIGHT_BOB).build();
+        assertFalse(ALICE.equals(editedAlice));
+
+        //different gender -> returns false
+        editedAlice = new ClientBuilder(ALICE).withGender("f").build();
```
Comment (jundatan) [inline src/test/java/seedu/address/model/client/ClientTest.java:96]: Missing Assert.

**Review by jundatan — CHANGES_REQUESTED** [review]
One missing assert. Everything else LGTM. Will merge after updating.

File: src/test/java/seedu/address/model/client/ClientTest.java (hunk 26 above)
Comment (dfordarius) [reply src/test/java/seedu/address/model/client/ClientTest.java:96]: oops good spot, thanks!

**Review by jundatan — APPROVED** [review]
